cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -402,16 +401,16 @@ struct StringTransformExecBase {
     if (!input.is_valid) {
       return Status::OK();
     }
-    auto* result = checked_cast<BaseBinaryScalar*>(out->scalar().get());
-    result->is_valid = true;
     const int64_t data_nbytes = static_cast<int64_t>(input.value->size());
-
     const int64_t output_ncodeunits_max = transform->MaxCodeunits(1, data_nbytes);
     if (output_ncodeunits_max > std::numeric_limits<offset_type>::max()) {
       return Status::CapacityError(
           "Result might not fit in a 32bit utf8 array, convert to large_utf8");
     }
+
     ARROW_ASSIGN_OR_RAISE(auto value_buffer, ctx->Allocate(output_ncodeunits_max));
+    auto* result = checked_cast<BaseBinaryScalar*>(out->scalar().get());
```

Comment:
This is a very good observation. A simple search through the C++ codebase shows that both patterns are used. I agree with having `nullptr` checks after `checked_cast()`. I will ask the in Zulip dev to see if this is a pattern we want to enforce. If so, then we should create JIRA.

Reply:
We could just include the check in checked_cast?

Reply:
After a bit of more thought, it is not that easy to enforce because in many cases the pointer is dereferenced beforehand:
```c++
const auto& obs = checked_cast<const Type&>(*some_var);
```
So the more general question is when should pointers  be checked for nullity? Should we check everywhere a raw pointer is accessed?

Reply:
Cases like `const auto& obs = checked_cast<const Type&>(*some_var);` will actually generate a `std::bad_cast` exception so we wouldn't have to worry about those cases.  I suspect it could be solved with specialization.  Zulip dev is probably a good place for it.  No need to tackle in this PR.